hadoop-hdfs-project/hadoop-hdfs/src/main/java/io/hops/transaction/context/FileProvenanceContext.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2019 hops.io.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.hops.transaction.context;
+
+import io.hops.exception.StorageException;
+import io.hops.exception.TransactionContextException;
+import io.hops.metadata.hdfs.entity.FileProvenanceEntry;
+import io.hops.transaction.lock.TransactionLocks;
+import java.util.Objects;
+import io.hops.metadata.hdfs.dal.FileProvenanceDataAccess;
+
+public class FileProvenanceContext extends BaseEntityContext<FileProvenanceContext.DBKey, FileProvenanceEntry> {
+  
+  private final FileProvenanceDataAccess<FileProvenanceEntry> dataAccess;
+
+  class DBKey {
+
+    private final long inodeId;
+    private final int userId;
+    private final String appId;
+    private final long timestamp;
+    private final long logicalTime;
+    private final String operation;
+    public DBKey(long inodeId, int userId, String appId, long timestamp, long logicalTime, String operation) {
+      this.inodeId = inodeId;
+      this.userId = userId;
+      this.appId = appId;
+      this.timestamp = timestamp;
+      this.logicalTime = logicalTime;
+      this.operation = operation;
+    }
+
+    @Override
+    public int hashCode() {
+      int hash = 3;
+      hash = 17 * hash + (int) (this.inodeId ^ (this.inodeId >>> 32));
+      hash = 17 * hash + this.userId;
+      hash = 17 * hash + Objects.hashCode(this.appId);
+      hash = 17 * hash + (int) (this.timestamp ^ (this.timestamp >>> 32));
+      return hash;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (this == obj) {
+        return true;
+      }
+      if (obj == null) {
+        return false;
+      }
+      if (getClass() != obj.getClass()) {
+        return false;
+      }
+      final DBKey other = (DBKey) obj;
+      if (this.inodeId != other.inodeId) {
+        return false;
+      }
+      if (this.userId != other.userId) {
+        return false;
+      }
+      if (this.timestamp != other.timestamp) {
+        return false;
+      }
+      if (!Objects.equals(this.appId, other.appId)) {
+        return false;
+      }
+      return true;
+    }
+  }
+
+  public FileProvenanceContext(FileProvenanceDataAccess dataAccess) {
+    this.dataAccess = dataAccess;
+  }
+
+  @Override
+  public void add(FileProvenanceEntry logEntry)
+    throws TransactionContextException {
+    DBKey dbKey = getKey(logEntry);
+    if (get(dbKey) != null) {
```

Comment:
As we never get from the database, what guaranty that there can't be a conflict with an opperation happening in parallel on another NN?

Reply:
The operations should be serialized per inode using the write lock where we increment the logical time. However, i still have the same concern since i can see at some part of the code that you also call log_provenance in the getBlockLocations which will usually run with a READ lock and will not update the logical time. Is that true? or you don't support for now?

Reply:
Yes, that is an issue. In order to fix it, I added a auto increment tie breaker field in the primary key:
https://github.com/hopshadoop/hops-metadata-dal-impl-ndb/pull/225/files#diff-337b19c10be62b3bf39c13b55a765697R18
with the comment:
https://github.com/hopshadoop/hops-metadata-dal-impl-ndb/pull/225/files#diff-8e6061b872fabdd5cc0806d04e11bfceR77
I want all these entries. The cases of actual interest are where the dataset provenance core xattr is changed concurrently with these reads. EPipe can then decide how to log these concurrent reads that might actually have a different provenance logging type(the dataset provenance core xattr was changed). If the dataset logical time was changed by an update to the core provenance xattr then epipe might have to see how to resolve this and how to log these apparent concurrent operations.

Reply:
mmm, do you depend on the monotonicity of this auto increment column to enforce ordering anywhere in HopsFS or ePipe? the autoincrement columns are not serialized by default therefore we will end up with gaps in the incremented ids. 
https://dev.mysql.com/doc/ndbapi/en/ch04s03s04s01.html
Default configuration is set to batch of 10 (DEFAULT_PROPERTY_CLUSTER_CONNECT_AUTO_INCREMENT_BATCH_SIZE)

Reply:
I will not use the auto increment field for total order... it is simply used for tie breaking so i get all the entries (even possible conflicts) in the log table.

Reply:
Ok, then it should be fine. However, we should be careful while using auto increment columns since they will slow down reads under load, maybe will require to increase the batch size.

Reply:
The main concern that auto increments in the critical path will be a bottleneck under load since it is a synchronization point even with a bigger batch.

Reply:
ok, as agreed, I removed the auto-increment column and replaced it with a namenodeId_threadId - identifier. This ids should break any ties between the read conflicts mentioned above